File: Lab4/Fraction.cpp
#include "Fraction.h"
#include <iostream>
#include <string>

using namespace std;

namespace Mathematics
{
	void Fraction::reduce(void)
	{
		if (!denominator)
			throw invalid_argument("Знаменатель дроби не должен быть равен 0. Деление на ноль\n");

		if (denominator < 0)
		{
			denominator = -denominator;
			numerator = -numerator;
		}

		int div = nod(numerator, denominator);

		numerator /= div;
		denominator /= div;

		if (!numerator)
			denominator = 1;
	}

	Fraction::Fraction(int _numerator, int _denominator)
	{
		numerator = _numerator;
		denominator = _denominator;

		reduce();
	}

	pair<int, Fraction> Fraction::toProper(void)
	{
		return pair<int, Fraction> {numerator / denominator, Fraction(numerator % denominator, denominator)};
	}

	Fraction & Fraction::toImproper(int _integerPart)
	{
		*this += _integerPart;

		return *this;
	}

	int Fraction::getNumerator(void) const
	{
		return numerator;
	}

	int Fraction::getDenominator(void) const
	{
		return denominator;
	}

	Fraction & Fraction::operator += (const Fraction &_a)
	{
		__int64 newDen = nok(denominator, _a.denominator);

		numerator = (int)(numerator * (newDen / denominator) + _a.numerator * (newDen / _a.denominator));
		denominator = (int)newDen;

		reduce();

		return *this;
	}
	
	Fraction & Fraction::operator -= (const Fraction &_a)
	{
		__int64 newDen = nok(denominator, _a.denominator);

		numerator = (int)(numerator * (newDen / denominator) - _a.numerator * (newDen / _a.denominator));
		denominator = (int)newDen;

		reduce();

		return *this;
	}
	
	Fraction & Fraction::operator *= (const Fraction &_a)
	{
		numerator *= _a.numerator;
		denominator *= _a.denominator;

		reduce();

		return *this;
	}

	Fraction & Fraction::operator /= (const Fraction &_a)
	{
		numerator *= _a.denominator;
		denominator *= _a.numerator;

		reduce();

		return *this;
	}

	bool operator == (const Fraction &_a, const Fraction & _b)
	{
		return _a.numerator == _b.numerator && _a.denominator == _b.denominator;
	}

	bool operator != (const Fraction & _a, const Fraction & _b)
	{
		return !(_a == _b);
	}

	bool operator!(const Fraction & _a)
	{
		return !(_a.numerator);
	}

	bool operator > (const Fraction &_a, const Fraction & _b)
	{
		return (__int64)_a.numerator * (__int64)_b.denominator > (__int64)_a.denominator * (__int64)_b.numerator;
	}

	bool operator >= (const Fraction &_a, const Fraction & _b)
	{
		return _a == _b || _a > _b;
	}

	bool operator < (const Fraction &_a, const Fraction & _b)
	{
		return (__int64)_a.numerator *(__int64)_b.denominator < (__int64)_a.denominator * (__int64)_b.numerator;
	}

	bool operator <= (const Fraction &_a, const Fraction & _b)
	{
		return _a == _b || _a < _b;
	}

	Fraction operator + (const Fraction &_a, const Fraction & _b)
	{
		Fraction c(_a);

		return c += _b;
	}

	Fraction operator - (const Fraction &_a, const Fraction & _b)
	{
		Fraction c(_a);

		return c -= _b;
	}

	Fraction operator * (const Fraction &_a, const Fraction & _b)
	{
		Fraction c(_a);

		return c *= _b;
	}

	Fraction operator / (const Fraction &_a, const Fraction & _b)
	{
		Fraction c(_a);

		return c /= _b;
	}

	ostream & operator << (ostream & _out, const Fraction &_a)
	{

		if (_a.denominator == 1)
			_out << _a.numerator;
		else
			_out << _a.numerator << "/" << _a.denominator;

		return _out;
	}

	int nok(int _firstEl, int _secondEl)
	{
		return (_firstEl * _secondEl / nod(_firstEl, _secondEl));
	}

	int nod(int _firstEl, int _secondEl)
	{
		if (_firstEl < 0)
			_firstEl = -_firstEl;

		if (_secondEl < 0)
			_secondEl = -_secondEl;

		for (int i = _firstEl; i > 0; i--)
		{
			if (_firstEl % i == 0 && _secondEl % i == 0)
			{
				return i;
			}
		}

		return 1;
	}

}

File: Lab8/Fraction.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Mathematics
{
	struct Fraction : IFormattable, ICloneable, IComparable, IComparable<Fraction>
	{
		private int denominator;
		private bool autoReduce;

		private void Reduce()
		{
			if (!AutoReduce)
				return;

			if (Denominator < 0)
			{
				denominator = -Denominator;
				Numerator = -Numerator;
			}

			int div = Nod(Numerator, Denominator);
			
			Numerator /= div;
			denominator /= div;

			if (Numerator == 0)
				denominator = 1;
		}

		public Fraction(int numerator, int denominator = 1, bool autoReduce = false)
		{
			if (denominator == 0)
				throw new DivideByZeroException("Знаменатель дроби не должен быть равен 0. Деление на ноль\n");

			Numerator = numerator;
			this.denominator = denominator;
			this.autoReduce = autoReduce;

			Reduce();
		}

		public Fraction(Fraction source) : this(source.Numerator, source.Denominator) { }

		public int ToProper()
		{
			Numerator = Numerator % Denominator;

			return Numerator / Denominator;
		}

		public static Fraction operator + (Fraction a, Fraction b)
		{
			int newDen = Nok(a.Denominator, b.Denominator);
		
			return new Fraction(a.Numerator * (newDen / a.Denominator) + b.Numerator * (newDen / b.Denominator), newDen, a.AutoReduce && b.AutoReduce);
		}

		public static Fraction operator - (Fraction a, Fraction b)
		{
			int newDen = Nok(a.Denominator, b.Denominator);

			return new Fraction(a.Numerator * (newDen / a.Denominator) - b.Numerator * (newDen / b.Denominator), newDen, a.AutoReduce && b.AutoReduce);
		}

		public static Fraction operator * (Fraction a, Fraction b) 
			=> new Fraction(a.Numerator * b.Numerator, a.Denominator * b.Denominator, a.AutoReduce && b.AutoReduce);

		public static Fraction operator / (Fraction a, Fraction b) 
			=> new Fraction(a.Numerator * b.Denominator, a.Denominator * b.Numerator, a.AutoReduce && b.AutoReduce);

		public static explicit operator int(Fraction a)
			=> a.Numerator / a.Denominator;

		public static explicit operator double(Fraction a)
			=> (double)a.Numerator / a.Denominator;

		public static implicit operator Fraction(int a)
			=> new Fraction(a);

		public int Numerator { get; private set; }

		public int Denominator
		{
			get
			{
				if (denominator == 0)
					denominator = 1;

				return denominator;
			}
		}

		public bool AutoReduce
		{
			get => autoReduce;

			set
			{
				autoReduce = value;

				Reduce();
			}
		}

		public override string ToString()
			=> ToString("G", CultureInfo.CurrentCulture);

		public string ToString(string format)
			=> ToString(format, CultureInfo.CurrentCulture);

		public string ToString(IFormatProvider provider)
			=> ToString("G", provider);

		public string ToString(string format, IFormatProvider provider)
		{
			provider = provider ?? CultureInfo.CurrentCulture;

			if (string.IsNullOrEmpty(format))
				format = "G";

			switch (format.ToUpperInvariant())
			{
				case "G":
					if (Denominator == 1)
						return string.Format(provider, "{0}", Numerator);
					else
						return string.Format(provider, "{0}/{1}", Numerator, Denominator);
				default:
					throw new FormatException(string.Format($"Ошибка! Формат {format} неизвестен."));
			}
		}

		public static int Nok(int a, int b)
			=> a * b / Nod(a, b);

		public static int Nod(int a, int b)
		{
			if (a < 0)
				a = -a;

			if (b < 0)
				b = -b;

			for (int i = a; i > 0; i--)
				if (a % i == 0 && b % i == 0)
					return i;

			return 1;
		}

		object ICloneable.Clone()
			=> Clone();
		
		public Fraction Clone()
			=> new Fraction(this);

		public static bool operator == (Fraction a, Fraction b)
			=> a.Numerator == b.Numerator && a.Denominator == b.Denominator;
		
		public static bool operator != (Fraction a, Fraction b)
			=> !(a == b);

		public static bool operator > (Fraction a, Fraction b)
			=> (long)a.Numerator * b.Denominator > (long)a.Denominator * b.Numerator;

		public static bool operator < (Fraction a, Fraction b)
			=> (long)a.Numerator * b.Denominator < (long)a.Denominator * b.Numerator;

		public static bool operator >= (Fraction a, Fraction b)
			=> a == b || a > b;

		public static bool operator <= (Fraction a, Fraction b)
			=> a == b || a < b;

		public override bool Equals(object obj)
			=> base.Equals(obj);

		public override int GetHashCode()
			=> Numerator * Denominator ^ 8;

		public int CompareTo(object obj)
		{
			if (!(obj is Fraction))
				throw new ArgumentException("Объект не совпадает с типом (Fraction) данного экземпляра.");

			return CompareTo((Fraction)obj);
		}

		public int CompareTo(Fraction other)
			=> this == other ? 0 : this > other ? 1 : -1;
	} 
}

